Add ecrecover alias map to evm contract

prestwich opened the pull request:
supersedes #28 by integrating the logic directly into the contract

See this comment for rationale: https://github.com/near/near-evm/pull/28#issuecomment-645544390

# TODOs
- [x] add method for registering alias
- [x] improve protocol binding
- [x] tests of registration and aliasing
- [ ] blocked by https://github.com/nearprotocol/nearcore/issues/2939

prestwich committed:
WIP(feature): add ecrecover alias map to evm contract

ewilz commented on src/builtins.rs:
```diff
@@ -44,6 +45,16 @@ pub fn process_precompile(addr: &Address, input: &[u8]) -> MessageCallResult {
         .expect("No errors in precompiles");
 
+    // hijack ECDSA calls
+    if addr.to_low_u64_be() == 1 {
```
I'm guessing this checks to make sure the address produced from ecrecover matches the address associated with the account, but really I'm confused by lines 51-54

prestwich replied:
We want to allow Near addresses to register ECDSA keys. so we need a way to make signatures under those keys appear as if they were a different account.

So line 51 looks for an alias using the `ecdsa_map_get` function. if it finds an alias (`if let Some(addr)`), it clears the current return value, and replaces it with the target of the alias

prestwich committed:
feature: address alias registration

prestwich committed:
docs: document register_ecdsa_alias

prestwich committed:
feature: improved protocol binding for ecrecover alias

prestwich committed:
docs: update docs for new ECDSA behavior

prestwich commented:
I've improved the protocol binding, and updated the documentation, and added simple tests for registration. Next major step is to add a test for the aliasing

prestwich committed:
tests: simple test for ecrecover aliasing effectiveness

prestwich committed:
cleanup: comment and borrow

ewilz commented on src/lib.rs:
```diff
@@ -354,6 +360,67 @@ impl EvmContract {
+    ///
+    /// * When `signature` is not valid hex.
+    /// * When the signature is invalid.
+    pub fn register_ecdsa_alias(&mut self, signature: String) {
```
can an address register multiple aliases?

prestwich replied:
yes, but there's currently no way to unregister them.

ewilz replied:
Also this may have to have a return value so web3 doesn't mistake it for a revert (see, for instance, line 318)

ewilz commented on src/evm_state.rs:
```diff
@@ -231,6 +237,10 @@ impl SubState<'_> {
 
 impl EvmState for SubState<'_> {
+    fn get_ecrecover_alias(&self, address: Address) -> Option<Address> {
+        self.parent.get_ecrecover_alias(address)
```
if parent is `impl EvmState for StateStore` will this return `None` or filter up somehow? I guess I need a refresher on these

prestwich committed:
refactor: docs and add bool return value to register_ecdsa_alias

prestwich committed:
feature: ECDSA alias revocation

prestwich committed:
feature: don't drop topics from logs

prestwich committed:
Merge pull request #44 from near/log-topics

Log topics

ewilz commented on src/near_ext.rs:
```diff
@@ -259,18 +259,26 @@ impl<'a> vm::Ext for NearExt<'a> {
 
     /// Creates log entry with given topics and data
-    fn log(&mut self, _topics: Vec<H256>, data: &[u8]) -> EvmResult<()> {
+    fn log(&mut self, topics: Vec<H256>, data: &[u8]) -> EvmResult<()> {
```
where does this function get called? I was looking at it the other day and couldn't find where

prestwich replied:
It's called by the parity evm interpreter whenever a log opcode is executed

ewilz commented:
is there a change near is making that will definitely fix up our capacity to deploy the evm?? I commented out all of builtin.rs except for the ecsda stuff and still couldn't deploy

prestwich commented:
Yeah, it's referenced in the PR description.

ewilz commented:
> Yeah, it's referenced in the PR description.

ok, cool. was just checking that it was outside the scope of the builtin code since that omission still didn't do the trick

ewilz commented on src/lib.rs:
```diff
@@ -419,12 +424,54 @@ impl EvmContract {
+        let alias = utils::hex_to_evm_address(&alias_address);
+
+        if self.get_ecrecover_alias(alias) == Some(sender) {
+            self.set_ecrecover_alias(&alias, &alias);
```
if evm alias == evm alias means it's revoked? no zero val? (nitpick)

prestwich replied:
I did this so that an alias can never be re-registered by another near account. Setting to 0 value would allow re-use of the alias in potentially unsafe ways

ewilz replied:
ahhh ok

ilblackdragon commented:
Outdated. Going to close.
